backup: always write to cloud storage in processor

dt opened the pull request:
Release note (sql change): the setting bulkio.backup.proxy_file_writes.enabled is no longer needed to enable proxied writes which are now the default.

Also, while I'm here, de-couple the target file size used by the processor (which streams its writes out to cloud storage) from the target file size of ExportRequest (which is now always in-memory, and twice over at that, on both the KV and SQL side of the RPC).

dt committed:
backupccl: use separate BACKUP and Export target file sizes

The BACKUP processor should pick its target file size -- at which it closes
a file and opens a new one -- independently from the KV's target Export file
size, which has factors like request duration, memory footprint, etc that may
influence it but are not relevant to the backup file as it is streamed our to
cloud storage.

Release note: none.

dt committed:
backup: always write to cloud storage in processor

This removes the bulkio.backup.proxy_file_writes.enabled setting, and just
always does the behavior of it being true.

Release note (sql change): the setting bulkio.backup.proxy_file_writes.enabled is no longer needed to enable proxied writes which are now the default.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68468)
<!-- Reviewable:end -->

dt commented:
@pbardea should we also drop `kv.bulk_sst.target_size` to 8 or 16mb while we're at it?

dt committed:
backup: reduce default kv.bulk_sst.target_size to 16mb

Release note: none.

pbardea commented:
> should we also drop kv.bulk_sst.target_size to 8 or 16mb while we're at it?

Yep, that sounds reasonable. I would think that 16 MB should be small enough for a sane default?

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68468#-:-MgLt6dJ7BNINIJA5uW9:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aliher1911)



<!-- Sent from Reviewable.io -->

dt committed:
backup: add some VEventfs to backup file flushing

Release note: none.

dt committed:
backup: fix bug when counting progress for returned SSTs

Release note: none.

dt commented:
Some revisions here that probably deserve a re-review (sorry!):
- I found an existing bug in progress reporting for returned SSTs fixed in what is now the first commit,
- added some extra debug VEventf's, and 
- ripped out support for _non_-returned ssts from the processor's ExportResponse handling

pbardea commented on pkg/ccl/backupccl/backup_processor.go:
```diff
@@ -422,47 +398,22 @@ func runBackupProcessor(
 						// If this file reply has an inline SST, push it to the
```
Let's update this comment too (the reply will always have an inline SST)

pbardea commented on pkg/ccl/backupccl/backup_processor.go:
```diff
@@ -422,47 +398,22 @@ func runBackupProcessor(
-							}
-						} else {
-							files = append(files, f)
+						exportResponseTraceEvent.HasReturnedSSTs = true
```
Is this field still worth keeping around in the trace event if it's always true?

pbardea commented on pkg/ccl/backupccl/backup_processor.go:
Can we simplify the condition below, as well as the comment on top of it (L304, but GH won't let me comment there):

```
if req.ReturnSST {
						header.TargetBytes = 1
}
```

pbardea reviewed (approved):
Nice bug spotting! LGTM with nits

dt committed:
backup: use processor's locality to pick backup partition

Locality-partitioned backups were intended to reduce cross-regional file writes.
Thus, if the processor is writing a file, it should do so to the region that matches
its locality. Due to PartitionSpans, this should generally be the same node and thus
same region as the leaseholder, or at least where the leaseholder was during planning.

Release note: none.

aliher1911 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68468#-:-MgQE-_wBjuHPF2BB0Px:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @dt and @nvanbenschoten)

___
*[pkg/ccl/backupccl/backup_processor.go, line 386 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68468#-MgQDNlP2XtpFAcOh3RE:-MgQDNlP2XtpFAcOh3RF:b-919mxx) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c5979105007c7e1845a09a6032b603077128391/pkg/ccl/backupccl/backup_processor.go#L386)):*
> ```Go
> 						FileSummaries: make([]RowCount, 0),
> 					}
> 					for i, file := range res.Files {
> ```

Maybe it is worth adding a comment that we always expect a single file here and loop is kept for historic reasons? I think there are few checks here like completedSpans below that give an impression that there could be more that a single entry so comment at least in one place could make it easier to understand.


<!-- Sent from Reviewable.io -->

dt committed:
backup: log warning on unexpected multi-file response

With header.TargetBytes = 1 we expect each request to return at most one file,
but the loop, kept here for legacy reasons may suggest otherwise. Logging a warning
should help document what we expect while also informing us if we're mistaken.

Release note: none.

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68468)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @aliher1911, @dt, and @nvanbenschoten)

___
*[pkg/ccl/backupccl/backup_processor.go, line 386 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68468#-MgQDNlP2XtpFAcOh3RE:-MgQgrM-0iZoeq9geiiS:b-qhhmy6) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c5979105007c7e1845a09a6032b603077128391/pkg/ccl/backupccl/backup_processor.go#L386)):*
<details><summary><i>Previously, aliher1911 (Oleg) wrote…</i></summary><blockquote>

Maybe it is worth adding a comment that we always expect a single file here and loop is kept for historic reasons? I think there are few checks here like completedSpans below that give an impression that there could be more that a single entry so comment at least in one place could make it easier to understand.
</blockquote></details>

added a logged warning if len > 1 (figure that both documents expectation as a comment would, but also informs us if they're incorrect?)


<!-- Sent from Reviewable.io -->

dt commented:
TFTRs!

bors r+
